src/libraries/System.Web.HttpUtility/src/System/Web/Util/HttpEncoder.cs

```diff
@@ -130,79 +123,69 @@ internal static void HtmlEncode(string? value, TextWriter output)
-                // to the string builder when special characters are detected.
-                if (CharRequiresJavaScriptEncoding(c))
-                {
-                    b ??= new StringBuilder(value.Length + 5);
-
-                    if (count > 0)
-                    {
-                        b.Append(value, startIndex, count);
-                    }
-
-                    startIndex = i + 1;
-                    count = 0;
-
-                    switch (c)
-                    {
-                        case '\r':
-                            b.Append("\\r");
-                            break;
-                        case '\t':
-                            b.Append("\\t");
-                            break;
-                        case '\"':
-                            b.Append("\\\"");
-                            break;
-                        case '\\':
-                            b.Append("\\\\");
-                            break;
-                        case '\n':
-                            b.Append("\\n");
-                            break;
-                        case '\b':
-                            b.Append("\\b");
-                            break;
-                        case '\f':
-                            b.Append("\\f");
-                            break;
-                        default:
-                            AppendCharAsUnicodeJavaScript(b, c);
-                            break;
-                    }
-                }
-                else
-                {
-                    count++;
-                }
+                return addDoubleQuotes ? $"\"{value}\"" : value;
             }
 
-            if (b == null)
+            var vsb = new ValueStringBuilder(stackalloc char[512]);
+            if (addDoubleQuotes)
             {
-                return value;
+                vsb.Append('"');
             }
+            ReadOnlySpan<char> chars = value;
+            do
+            {
+                vsb.Append(chars.Slice(0, i));
+                char c = chars[i];
+                chars = chars.Slice(i + 1);
+                switch (c)
+                {
+                    case '\r':
+                        vsb.Append("\\r");
+                        break;
+                    case '\t':
+                        vsb.Append("\\t");
+                        break;
+                    case '\"':
+                        vsb.Append("\\\"");
+                        break;
+                    case '\\':
+                        vsb.Append("\\\\");
+                        break;
+                    case '\n':
+                        vsb.Append("\\n");
+                        break;
+                    case '\b':
+                        vsb.Append("\\b");
+                        break;
+                    case '\f':
+                        vsb.Append("\\f");
+                        break;
+                    default:
+                        vsb.Append($"\\u{(int)c:x4}");
```

Comment:
```suggestion
                        vsb.Append("\\u");
                        vsb.AppendSpanFormattable((int)c, "x4");
```

Unlike `StringBuilder`, `VSB` doesn't have an append overload with an interpolated string handler.
This avoids the temporary string allocation with VSB as well.